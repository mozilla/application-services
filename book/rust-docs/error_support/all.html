<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><title>List of all items in this crate</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="error_support" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0 (129f3b996 2024-06-10)" data-channel="1.79.0" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod sys"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../error_support/index.html">error_support</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><section><ul class="block"><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../error_support/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><h1>List of all items</h1><h3 id="structs">Structs</h3><ul class="all-items"><li><a href="struct.ErrorHandling.html">ErrorHandling</a></li><li><a href="struct.ErrorReporting.html">ErrorReporting</a></li></ul><h3 id="traits">Traits</h3><ul class="all-items"><li><a href="trait.ApplicationErrorReporter.html">ApplicationErrorReporter</a></li><li><a href="trait.GetErrorHandling.html">GetErrorHandling</a></li></ul><h3 id="macros">Macros</h3><ul class="all-items"><li><a href="macro.breadcrumb.html">breadcrumb</a></li><li><a href="macro.define_error.html">define_error</a></li><li><a href="macro.define_error_conversions.html">define_error_conversions</a></li><li><a href="macro.define_error_wrapper.html">define_error_wrapper</a></li><li><a href="macro.report_error.html">report_error</a></li><li><a href="macro.trace_error.html">trace_error</a></li></ul><h3 id="attributes">Attribute Macros</h3><ul class="all-items"><li><a href="attr.handle_error.html">handle_error</a></li></ul><h3 id="functions">Functions</h3><ul class="all-items"><li><a href="fn.convert_log_report_error.html">convert_log_report_error</a></li><li><a href="fn.redact_compact_jwe.html">redact_compact_jwe</a></li><li><a href="fn.redact_url.html">redact_url</a></li><li><a href="fn.report_breadcrumb.html">report_breadcrumb</a></li><li><a href="fn.report_error_to_app.html">report_error_to_app</a></li><li><a href="fn.set_application_error_reporter.html">set_application_error_reporter</a></li><li><a href="fn.unset_application_error_reporter.html">unset_application_error_reporter</a></li></ul></section></div></main></body></html>