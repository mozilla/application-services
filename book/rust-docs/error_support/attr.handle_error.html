<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="A procedural macro that exposes internal errors to external errors the consuming applications should handle. It requires that the internal error implements [`error_support::ErrorHandling`]."><title>handle_error in error_support - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-916cea96.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="error_support" data-themes="" data-resource-suffix="" data-rustdoc-version="1.87.0 (17067e9ac 2025-05-09)" data-channel="1.87.0" data-search-js="search-e7298875.js" data-settings-js="settings-d72f25bb.js" ><script src="../static.files/storage-82c7156e.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-fb8c74a8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc attr"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../error_support/index.html">error_<wbr>support</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">handle_<wbr>error</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#example" title="Example">Example</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="index.html">In crate error_<wbr>support</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="index.html">error_support</a></div><h1>Attribute Macro <span class="attr">handle_error</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/error_support_macros/lib.rs.html#63">Source</a> </span></div><pre class="rust item-decl"><code>#[handle_error]</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>A procedural macro that exposes internal errors to external errors the
consuming applications should handle. It requires that the internal error
implements [<code>error_support::ErrorHandling</code>].</p>
<p>Additionally, this procedural macro has side effects, including:</p>
<ul>
<li>It would log the error based on a pre-defined log level. The log level is defined
in the [<code>error_support::ErrorHandling</code>] implementation.</li>
<li>It would report some errors using an external error reporter, in practice, this
is implemented using Sentry in the app.</li>
</ul>
<h2 id="example"><a class="doc-anchor" href="#example">§</a>Example</h2>
<div class="example-wrap ignore"><a href="#" class="tooltip" title="This example is not tested">ⓘ</a><pre class="rust rust-example-rendered"><code><span class="kw">use </span>error_support::{handle_error, GetErrorHandling, ErrorHandling};
<span class="kw">use </span>std::fmt::Display
<span class="attr">#[derive(Debug, thiserror::Error)]
</span><span class="kw">struct </span>Error {}
<span class="kw">type </span><span class="prelude-ty">Result</span>&lt;T, E = Error&gt; = std::result::Result&lt;T, E&gt;;

<span class="kw">impl </span>Display <span class="kw">for </span>Error {
    <span class="kw">fn </span>fmt(<span class="kw-2">&amp;</span><span class="self">self</span>, f: <span class="kw-2">&amp;mut </span>std::fmt::Formatter&lt;<span class="lifetime">'_</span>&gt;) -&gt; std::fmt::Result {
        <span class="macro">write!</span>(f, <span class="string">"Internal Error!"</span>)
    }
}

<span class="attr">#[derive(Debug, thiserror::Error)]
</span><span class="kw">struct </span>ExternalError {}

<span class="kw">impl </span>Display <span class="kw">for </span>ExternalError {
    <span class="kw">fn </span>fmt(<span class="kw-2">&amp;</span><span class="self">self</span>, f: <span class="kw-2">&amp;mut </span>std::fmt::Formatter&lt;<span class="lifetime">'_</span>&gt;) -&gt; std::fmt::Result {
        <span class="macro">write!</span>(f, <span class="string">"External Error!"</span>)
    }
}

<span class="kw">impl </span>GetErrorHandling <span class="kw">for </span>Error {
   <span class="kw">type </span>ExternalError = ExternalError;

   <span class="kw">fn </span>get_error_handling(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; ErrorHandling&lt;<span class="self">Self</span>::ExternalError&gt; {
       ErrorHandling::convert(ExternalError {})
   }
}

<span class="comment">// The `handle_error` macro maps from the error supplied in the mandatory argument
// (ie, `Error` in this example) to the error returned by the function (`ExternalError`
// in this example)
</span><span class="attr">#[handle_error(Error)]
</span><span class="kw">fn </span>do_something() -&gt; std::result::Result&lt;String, ExternalError&gt; {
   <span class="prelude-val">Err</span>(Error{})
}

<span class="comment">// The error here is an `ExternalError`
</span><span class="kw">let _</span>: ExternalError = do_something().unwrap_err();</code></pre></div>
</div></details></section></div></main></body></html>