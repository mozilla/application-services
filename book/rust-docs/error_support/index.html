<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `error_support` crate."><title>error_support - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="error_support" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0 (129f3b996 2024-06-10)" data-channel="1.79.0" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../error_support/index.html">error_support</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../error_support/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">error_support</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/error_support/lib.rs.html#5-162">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.backtrace"><code>pub use <a class="mod" href="https://docs.rs/backtrace/backtrace/index.html" title="mod backtrace">backtrace</a>;</code></div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.breadcrumb.html" title="macro error_support::breadcrumb">breadcrumb</a></div><div class="desc docblock-short">Tell the application to log a breadcrumb</div></li><li><div class="item-name"><a class="macro" href="macro.define_error.html" title="macro error_support::define_error">define_error</a></div><div class="desc docblock-short">All the error boilerplate (okay, with a couple exceptions in some cases) in
one place.</div></li><li><div class="item-name"><a class="macro" href="macro.define_error_conversions.html" title="macro error_support::define_error_conversions">define_error_conversions</a></div><div class="desc docblock-short">Define a set of conversions from external error types into the provided
error kind. Use <code>define_error</code> to do this at the same time as
<code>define_error_wrapper</code>.</div></li><li><div class="item-name"><a class="macro" href="macro.define_error_wrapper.html" title="macro error_support::define_error_wrapper">define_error_wrapper</a></div><div class="desc docblock-short">XXX - Most of this is now considered deprecated - only FxA uses it, and
should be replaced with the facilities in the <code>handling</code> module.
Define a wrapper around the the provided ErrorKind type.
See also <code>define_error</code> which is more likely to be what you want.</div></li><li><div class="item-name"><a class="macro" href="macro.report_error.html" title="macro error_support::report_error">report_error</a></div><div class="desc docblock-short">Tell the application to report an error</div></li><li><div class="item-name"><a class="macro" href="macro.trace_error.html" title="macro error_support::trace_error">trace_error</a></div><div class="desc docblock-short">Log a breadcrumb if we see an <code>Result::Err</code> value</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ErrorHandling.html" title="struct error_support::ErrorHandling">ErrorHandling</a></div><div class="desc docblock-short">Specifies how an “internal” error is converted to an “external” public error and
any logging or reporting that should happen.</div></li><li><div class="item-name"><a class="struct" href="struct.ErrorReporting.html" title="struct error_support::ErrorReporting">ErrorReporting</a></div><div class="desc docblock-short">Describes what error reporting action should be taken.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.ApplicationErrorReporter.html" title="trait error_support::ApplicationErrorReporter">ApplicationErrorReporter</a></div><div class="desc docblock-short">Application error reporting trait</div></li><li><div class="item-name"><a class="trait" href="trait.GetErrorHandling.html" title="trait error_support::GetErrorHandling">GetErrorHandling</a></div><div class="desc docblock-short">A trait to define how errors are converted and reported.</div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.convert_log_report_error.html" title="fn error_support::convert_log_report_error">convert_log_report_error</a></div><div class="desc docblock-short">Handle the specified “internal” error, taking any logging or error
reporting actions and converting the error to the public error.
Called by our <code>handle_error</code> macro so needs to be public.</div></li><li><div class="item-name"><a class="fn" href="fn.redact_compact_jwe.html" title="fn error_support::redact_compact_jwe">redact_compact_jwe</a></div><div class="desc docblock-short">Redact compact jwe string (Five base64 segments, separated by <code>.</code> chars)</div></li><li><div class="item-name"><a class="fn" href="fn.redact_url.html" title="fn error_support::redact_url">redact_url</a></div><div class="desc docblock-short">Redact a URL.</div></li><li><div class="item-name"><a class="fn" href="fn.report_breadcrumb.html" title="fn error_support::report_breadcrumb">report_breadcrumb</a></div></li><li><div class="item-name"><a class="fn" href="fn.report_error_to_app.html" title="fn error_support::report_error_to_app">report_error_to_app</a></div></li><li><div class="item-name"><a class="fn" href="fn.set_application_error_reporter.html" title="fn error_support::set_application_error_reporter">set_application_error_reporter</a></div></li><li><div class="item-name"><a class="fn" href="fn.unset_application_error_reporter.html" title="fn error_support::unset_application_error_reporter">unset_application_error_reporter</a></div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.handle_error.html" title="attr error_support::handle_error">handle_error</a></div><div class="desc docblock-short">A procedural macro that exposes internal errors to external errors the
consuming applications should handle. It requires that the internal error
implements [<code>error_support::ErrorHandling</code>].</div></li></ul></section></div></main></body></html>