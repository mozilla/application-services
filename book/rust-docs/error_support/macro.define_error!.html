<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.define_error.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.define_error.html">macro.define_error.html</a>...</p>
    <script>location.replace("macro.define_error.html" + location.search + location.hash);</script>
</body>
</html>