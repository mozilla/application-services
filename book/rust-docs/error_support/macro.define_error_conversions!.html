<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.define_error_conversions.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.define_error_conversions.html">macro.define_error_conversions.html</a>...</p>
    <script>location.replace("macro.define_error_conversions.html" + location.search + location.hash);</script>
</body>
</html>