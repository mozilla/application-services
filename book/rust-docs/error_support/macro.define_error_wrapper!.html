<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.define_error_wrapper.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.define_error_wrapper.html">macro.define_error_wrapper.html</a>...</p>
    <script>location.replace("macro.define_error_wrapper.html" + location.search + location.hash);</script>
</body>
</html>