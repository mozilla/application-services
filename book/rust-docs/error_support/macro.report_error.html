<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Tell the application to report an error"><title>report_error in error_support - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-916cea96.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="error_support" data-themes="" data-resource-suffix="" data-rustdoc-version="1.87.0 (17067e9ac 2025-05-09)" data-channel="1.87.0" data-search-js="search-e7298875.js" data-settings-js="settings-d72f25bb.js" ><script src="../static.files/storage-82c7156e.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-fb8c74a8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../error_support/index.html">error_<wbr>support</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="index.html">error_support</a></div><h1>Macro <span class="macro">report_error</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/error_support/macros.rs.html#20-26">Source</a> </span></div><pre class="rust item-decl"><code>macro_rules! report_error {
    ($type_name:expr, $($arg:tt)*) =&gt; { ... };
}</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Tell the application to report an error</p>
<p>If configured by the application, this sent to the application, which should report it to a
Sentry-like system. This should only be used for errors that we donâ€™t expect to see and will
work on fixing if we see a non-trivial volume of them.</p>
<p>type_name identifies the error.  It should be the main text that gets shown to the
user and also how the error system groups errors together.  It should be in UpperCamelCase
form.</p>
<p>Good type_names require some trial and error, for example:</p>
<ul>
<li>Start with the error kind variant name</li>
<li>Add more text to distinguish errors more.  For example an error code, or an extra word
based on inspecting the error details</li>
</ul>
</div></details></section></div></main></body></html>