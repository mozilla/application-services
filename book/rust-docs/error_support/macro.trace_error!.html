<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.trace_error.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.trace_error.html">macro.trace_error.html</a>...</p>
    <script>location.replace("macro.trace_error.html" + location.search + location.hash);</script>
</body>
</html>