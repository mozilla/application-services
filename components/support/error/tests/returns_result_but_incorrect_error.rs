/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

 use std::fmt::Display;

 use error_support::{handle_error, ErrorHandling, GetErrorHandling};
 
 #[derive(Debug, thiserror::Error)]
 struct Error {}
 impl Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Internal Error!")
    }
}

 type Result<T, E = Error> = std::result::Result<T, E>;

 
 
 #[derive(Debug, thiserror::Error)]
 struct ExternalError {}

 #[derive(Debug, thiserror::Error)]
 struct OtherExternalError {}

 impl Display for ExternalError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         Ok(())
     }
 }


 impl Display for OtherExternalError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Ok(())
    }
}
 
 impl GetErrorHandling for Error {
     type ExternalError = OtherExternalError;
 
     fn get_error_handling(&self) -> error_support::ErrorHandling<Self::ExternalError> {
         ErrorHandling::convert(OtherExternalError {})
     }
 }

 // handle_error expects that `Error` implements GetErrorHandling<E = ExternalError>
 // instead, it implements GetErrorHandling<E = OtherExternalError>
 #[handle_error(Error)]
 fn func() -> Result<String, ExternalError> {
     Err(Error{})
 }

 fn main() {}
