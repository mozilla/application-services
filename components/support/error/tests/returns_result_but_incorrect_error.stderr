error[E0308]: mismatched types
  --> returns_result_but_incorrect_error.rs:50:2
   |
50 |  #[handle_error(Error)]
   |  ^^^^^^^^^^^^^^^^^^^^^^ expected `Result<String, ExternalError>`, found `Result<_, OtherExternalError>`
51 |  fn func() -> Result<String, ExternalError> {
   |               ----------------------------- expected `std::result::Result<String, ExternalError>` because of return type
   |
   = note: expected enum `std::result::Result<String, ExternalError>`
              found enum `std::result::Result<_, OtherExternalError>`
   = note: this error originates in the attribute macro `handle_error` (in Nightly builds, run with -Z macro-backtrace for more info)
